src/mono/netcore/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.Mono.cs

```diff
@@ -354,6 +356,85 @@ public static unsafe IntPtr StringToBSTR(string? s)
                 return BufferToBSTR(fixed_s, s.Length);
         }
 
+        internal class MarshalerInstanceKeyComparer : IEqualityComparer<(Type, string)>
+        {
+            public bool Equals((Type, string) lhs, (Type, string) rhs)
+            {
+                return lhs.CompareTo(rhs) == 0;
+            }
+
+            public int GetHashCode((Type, string) key)
+            {
+                return key.GetHashCode();
+            }
+        }
+
+        internal static Dictionary<(Type, string), ICustomMarshaler>? MarshalerInstanceCache;
+
+        internal static ICustomMarshaler? GetCustomMarshalerInstance(Type type, string cookie)
+        {
+            var key = (type, cookie);
+
+            LazyInitializer.EnsureInitialized(
+                ref MarshalerInstanceCache,
+                () => new Dictionary<(Type, string), ICustomMarshaler>(new MarshalerInstanceKeyComparer())
+            );
+
+            ICustomMarshaler? result;
+            bool gotExistingInstance;
+            lock (MarshalerInstanceCache)
+                gotExistingInstance = MarshalerInstanceCache.TryGetValue(key, out result);
+
+            if (!gotExistingInstance)
+            {
+                RuntimeMethodInfo? getInstanceMethod;
+                try
+                {
+                    getInstanceMethod = (RuntimeMethodInfo?)type.GetMethod(
+                        "GetInstance", BindingFlags.Static | BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.InvokeMethod,
+                        null, new Type[] { typeof(string) }, null
+                    );
+                }
+                catch (AmbiguousMatchException)
+                {
+                    throw new ApplicationException($"Custom marshaler '{type.FullName}' implements multiple static GetInstance methods that take a single string parameter.");
+                }
+
+                if ((getInstanceMethod == null) ||
+                    (getInstanceMethod.ReturnType != typeof(ICustomMarshaler)))
+                {
+                    throw new ApplicationException($"Custom marshaler '{type.FullName}' does not implement a static GetInstance method that takes a single string parameter and returns an ICustomMarshaler.");
+                }
+
+                Exception? exc;
+                try
+                {
+                    result = (ICustomMarshaler?)getInstanceMethod.InternalInvoke(null, new object[] { cookie }, out exc);
+                }
+                catch (Exception e)
+                {
+                    // FIXME: mscorlib's legacyUnhandledExceptionPolicy is apparently 1,
```

Comment:
Is this still accurate?

Reply:
CoreCLR doesn't seem to use this, but some of our tests' .config files are. At least it seems benign to me to have this catch block.